libraries/AP_NMEA_Output/AP_NMEA_Output.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+   Author: Francisco Ferreira (some code is copied from sitl_gps.cpp)
+
+ */
+
+#include "AP_NMEA_Output.h"
+
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_RTC/AP_RTC.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+
+#include <cstdio>
+
+AP_NMEA_Output* AP_NMEA_Output::_singleton;
+
+AP_NMEA_Output::AP_NMEA_Output()
+{
+    AP_SerialManager& sm = AP::serialmanager();
+
+    for (uint8_t i = 0; i < ARRAY_SIZE(_uart); i++) {
+        if ((_uart[i] = sm.find_serial(AP_SerialManager::SerialProtocol_NMEAOutput, i)) == nullptr) {
+            break;
+        }
+        _num_outputs++;
+    }
+}
+
+AP_NMEA_Output* AP_NMEA_Output::probe()
+{
+    if (!_singleton && AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_NMEAOutput, 0) != nullptr) {
+       _singleton = new AP_NMEA_Output();
+    }
+
+    return _singleton;
+}
+
+uint8_t AP_NMEA_Output::_nmea_checksum(const char *str)
+{
+    uint8_t checksum = 0;
+    const uint8_t* bytes = (const uint8_t*) str;
+
+    for (uint16_t i = 1; str[i]; i++) {
+        checksum ^= bytes[i];
+    }
+
+    return checksum;
+}
+
+void AP_NMEA_Output::update()
+{
+    uint16_t now = AP_HAL::millis16();
+    uint16_t time_diff = now - _last_sent;
+
+    // only send at 10Hz
+    if (time_diff < 100) {
+        return;
+    }
+
+    // get time and date
+    uint64_t time_usec;
+    bool time_valid = AP::rtc().get_utc_usec(time_usec);
+
+    if (!time_valid) {
+        return;
+    }
+
+    // not completely accurate, our time includes leap seconds and time_t should be without
+    time_t time_sec = time_usec / 1000000;
+    struct tm* tm = gmtime(&time_sec);
+
+    // format time string
+    char tstring[11];
+    snprintf(tstring, sizeof(tstring), "%02u%02u%06.3f", tm->tm_hour, tm->tm_min, tm->tm_sec + (time_usec % 1000000) * 1.0e-6);
+
+    // format date string
+    char dstring[7];
+    snprintf(dstring, sizeof(dstring), "%02u%02u%02u", tm->tm_mday, tm->tm_mon+1, tm->tm_year % 100);
+
+    AP_AHRS_NavEKF& ahrs = AP::ahrs_navekf();
+
+    // get location (note: get_position from AHRS always returns true after having GPS position once)
+    Location loc{};
+    bool pos_valid = ahrs.get_location(loc);
+
+    // format latitude
+    char lat_string[13];
+    double deg = fabs(loc.lat * 1.0e-7f);
```

Comment:
Why double?

Reply:
Because it is going to be used in a printf, so it needs to be a double anyway.

Reply:
still should be float, for the fabsf(). We don't want to link in fabs()